test/span.js

```diff
@@ -185,31 +184,80 @@ describe('span should', () => {
+                original: {
+                    operationName: originalOperation,
+                    samplingDecision: originalSamplingDecision,
+                },
+                new: {
+                    operationName: newOperation,
+                    samplingDecision: newSamplingDecision,
+                },
+                finalSamplingDecision: finalSamplingDecision
+            }
+        }
+        const testCases = [
+            makeTestCase(0, true, true, "idli", null, null, true),
+            makeTestCase(0, true, false, "idli", null, null, false),
+            makeTestCase(0, true, true, "poriyal", true, "murungakkai", true),
+            // We ignore the following case because setOperationName can only go from a sampling decision of 0->1
+            // makeTestCase(0, true, true, "poriyal", false, "murungakkai", false),
+            makeTestCase(0, true, false, "poriyal", true, "murungakkai", true),
+            makeTestCase(0, true, false, "poriyal", false, "murungakkai", false),
+
+            // Never call the sampler when the span is debug or sampled
+            makeTestCase(constants.SAMPLED_MASK, false, true, "avial", null, null, true),
+            makeTestCase(constants.SAMPLED_MASK, false, false, "avial", null, null, true),
+            makeTestCase(constants.SAMPLED_MASK, false, true, "avial", true, "puri", true),
+            makeTestCase(constants.SAMPLED_MASK, false, true, "avial", false, "puri", true),
+            makeTestCase(constants.SAMPLED_MASK, false, false, "avial", true, "puri", true),
+            makeTestCase(constants.SAMPLED_MASK, false, false, "avial", false, "puri", true),
+            
+            makeTestCase(constants.DEBUG_MASK, false, true, "avial", null, null, true),
+            makeTestCase(constants.DEBUG_MASK, false, false, "avial", null, null, true),
+            makeTestCase(constants.DEBUG_MASK, false, true, "avial", true, "puri", true),
+            makeTestCase(constants.DEBUG_MASK, false, true, "avial", false, "puri", true),
+            makeTestCase(constants.DEBUG_MASK, false, false, "avial", true, "puri", true),
+            makeTestCase(constants.DEBUG_MASK, false, false, "avial", false, "puri", true),
+            
+            makeTestCase(constants.SAMPLED_MASK | constants.DEBUG_MASK, false, true, "avial", null, null, true),
+            makeTestCase(constants.SAMPLED_MASK | constants.DEBUG_MASK, false, false, "avial", null, null, true),
+            makeTestCase(constants.SAMPLED_MASK | constants.DEBUG_MASK, false, true, "avial", true, "puri", true),
+            makeTestCase(constants.SAMPLED_MASK | constants.DEBUG_MASK, false, true, "avial", false, "puri", true),
+            makeTestCase(constants.SAMPLED_MASK | constants.DEBUG_MASK, false, false, "avial", true, "puri", true),
+            makeTestCase(constants.SAMPLED_MASK | constants.DEBUG_MASK, false, false, "avial", false, "puri", true),
```

Comment:
given the amount of tests, I would suggest moving this to another file `tests/upsampling.js`